nncf/torch/dynamic_graph/context.py

```diff
@@ -34,7 +34,12 @@
 from nncf.torch.dynamic_graph.trace_tensor import TensorMeta
 from nncf.common.utils.patcher import PATCHER
 
-_CURRENT_CONTEXT = None
+class ThreadLocalGlobalContext(threading.local):
+    def __init__(self):
+        super().__init__()
+        self.context = None
+
+_CURRENT_CONTEXT = ThreadLocalGlobalContext()
```

Comment:
While working on this PR, I fixed an obscure bug, making the NNCFNetwork actually thread-safe. Before the fix, `DataParallel` only worked due to sheer luck.